keps/sig-api-machinery/1965-kube-apiserver-identity/README.md

```diff
@@ -0,0 +1,281 @@
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+In a HA cluster, each kube-apiserver has an ID. Controllers have access to the
+list of IDs for living kube-apiservers in the cluster.
+
+## Motivation
+
+The [dynamic coordinated storage version API](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190802-dynamic-coordinated-storage-version.md#curating-a-list-of-participating-api-servers-in-ha-master)
+needs such a list to garbage collect stale records. The
+[API priority and fairness feature](https://github.com/kubernetes/kubernetes/pull/91389)
+needs a unique identifier for an apiserver reporting its concurrency limit.
+
+Currently, such a list is already maintained in the “kubernetes” endpoints,
+where the kube-apiservers’ advertised IP addresses are the IDs. However it is
+not working in all flavors of Kubernetes deployments. For example, if there is a
+load balancer for the cluster, where the advertise IP address is set to the IP
+address of the load balancer, all three kube-apiservers will have the same
+advertise IP address.
+
+## Proposal
+
+We will use “hostname+PID+random suffix (e.g. 6 base58 digits)” as the ID.
+
+Similar to the [node heartbeat](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0009-node-heartbeat.md),
+a kube-apiserver will store its ID in a Lease object. All kube-apiserver Leases
+will be stored in a special namespace “kube-apiserver-lease”. A controller will
+garbage collect expired Leases.
+
+### Caveats
+
+In this proposal we focus on kube-apiservers. Aggregated apiservers don’t have
+the same problem, because their record is already exposed via the service. By
+listing the pods selected by the service, an aggregated server can learn the
+list of living servers with distinct podIPs. A server can get its own IDs via
+downward API.
+
+We prefer false positives over false negatives, because false negatives are more
+harmful. In the storage version API scenario, if a kube-apiserver accidentally
+missed a heartbeat and got its Lease garbage collected, its StorageVersion can
+be falsely garbage collected as a consequence. In this case, the storage
+migrator won’t be able to migrate the storage, unless this kube-aipserver gets
+restarted and re-registers its StorageVersion. On the other hand, if a
+kube-apiserver is gone and its Lease still stays around for an hour or two, it
+will only delay the storage migration for the same period of time.
+
+### Risks and Mitigations
+
+A new namespace will be reserved for storing kube-apiserver identity Lease
```

Comment:
dig through the the community api and naming guidelines.  I'm pretty sure I reserved all of kube-*.

EDIT: found it https://github.com/kubernetes/community/blob/243f9930949f06f5104a7330c9d05993c4598ab3/contributors/devel/sig-architecture/api-conventions.md#namespace-names

Reply:
even got it into real docs https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#working-with-namespaces

Reply:
wow cool. Thanks for looking into it. I removed this risk.